### PR TITLE
Capture delivery outcomes

### PR DESCRIPTION
**Pull request by minglis**
Adds notification stats update into the callback process
- when a provider callback occurs and we update the status of the notification, also update the statistics table

Adds:
- Mapping object to the clients to handle mapping to various states from the response codes, this replaces the map.
- query lookup in the DAO to get the query based on response type / template type

Tests around rest class and dao to check correct updating of stats

Missing:
- multiple client callbacks will keep incrementing the counts of success/failure. This edge case needs to be handle in a future story.

### CODE REVIEW

File: app/clients/__init__.py [hunk 1]
```diff
@@ -11,3 +11,28 @@ class Client(object):
...
     '''
     pass
+
+
+STATISTICS_REQUESTED = 'requested'
+STATISTICS_DELIVERED = 'delivered'
+STATISTICS_FAILURE = 'failure'
+
+
+class ClientResponse:
```
Comment (minglis) [inline app/clients/__init__.py]: Base response class.

- Used to map status codes from SMS/Email providers back into Notify statuses

File: app/clients/email/aws_ses.py [hunk 2]
```diff
@@ -1,22 +1,33 @@
...
-        "success": True,
-        "notify_status": 'delivered'
-    },
-    'Complaint': {
-        "success": False,
-        "notify_status": 'complaint'
-    }
-}
+
+class AwsSesResponses(ClientResponse):
```
Comment (minglis) [inline app/clients/email/aws_ses.py]: SES Version of status mapper

File: app/clients/sms/firetext.py [hunk 3]
```diff
@@ -6,26 +6,34 @@
...
-        "notify_status": 'failed'
-    },
-    '2': {
-        "firetext_message": 'Undelivered (Pending with Network)',
-        "success": False,
-        "notify_status": 'sent'
-    }
-}
+
+class FiretextResponses(ClientResponse):
```
Comment (minglis) [inline app/clients/sms/firetext.py]: Firetext verison. Note here the "pending" state doesn't update the stats table (None state)

File: app/dao/notifications_dao.py [hunk 4]
```diff
@@ -49,20 +55,22 @@ def dao_create_notification(notification, notification_type):
...
 
-def update_notification_stats(notification, notification_type):
-    if notification_type == TEMPLATE_TYPE_SMS:
-        update = {
-            NotificationStatistics.sms_requested: NotificationStatistics.sms_requested + 1
-        }
-    else:
-        update = {
-            NotificationStatistics.emails_requested: NotificationStatistics.emails_requested + 1
+def update_query(notification_type, status):
```
Comment (minglis) [inline app/dao/notifications_dao.py]: Update queries mapped by type/outcome. Methods use notification type/status to pick correct update

File: app/notifications/rest.py [hunk 5]
```diff
@@ -11,8 +11,8 @@
 )
```
Comment (minglis) [inline app/notifications/rest.py]: Changes in this file are about using the new response objects defined in the clients.

**Comment by servingUpAces** [comment]
:+1:
